src/main/cpp/src/map_zip_with_utils.cu

```diff
@@ -0,0 +1,308 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "map_zip_with_utils.hpp"
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/labeling/label_segments.cuh>
+#include <cudf/lists/count_elements.hpp>
+#include <cudf/lists/gather.hpp>
+#include <cudf/lists/set_operations.hpp>
+#include <cudf/reduction.hpp>
+#include <cudf/table/experimental/row_operators.cuh>
+#include <cudf/table/table_view.hpp>
+#include <cudf/utilities/span.hpp>
+
+#include <thrust/scan.h>
+
+using namespace cudf;
+
+namespace spark_rapids_jni {
+
+[[maybe_unused]] std::unique_ptr<column> generate_labels(lists_column_view const& input,
+                                                         size_type n_elements,
+                                                         rmm::cuda_stream_view stream,
+                                                         rmm::device_async_resource_ref mr)
+{
+  auto labels = make_numeric_column(
+    data_type(type_to_id<size_type>()), n_elements, cudf::mask_state::UNALLOCATED, stream, mr);
+  auto const labels_begin = labels->mutable_view().template begin<size_type>();
+  cudf::detail::label_segments(
+    input.offsets_begin(), input.offsets_end(), labels_begin, labels_begin + n_elements, stream);
+  return labels;
+}
+
+[[maybe_unused]] std::unique_ptr<column> indices_of(
+  lists_column_view const& search_keys,    // Column containing lists of keys to search with
+  lists_column_view const& search_values,  // Column containing lists of values to search for
+  rmm::cuda_stream_view stream,            // CUDA stream for asynchronous execution
+  rmm::device_async_resource_ref mr)       // Memory resource for allocations
+{
+  CUDF_EXPECTS(search_keys.size() == search_values.size(),
+               "Number of search keys lists must match search values lists.");
+
+  // Get the number of lists (rows) in the input columns
+  auto const num_lists = search_values.size();
+
+  // Extract the child columns containing the actual key and value data
+  auto const all_keys   = search_keys.child();
+  auto const num_keys   = all_keys.size();
+  auto const all_values = search_values.child();
+
+  // Calculate the number of elements in each list for both keys and values
+  auto list_sizes = cudf::lists::count_elements(search_values, stream, mr);
```

Comment:
I had to look further down to understand which "list" this is a reference to.

This should mention `search_space` in the variable name.